### PR TITLE
hif: Initial comps group support.

### PR DESCRIPTION
**Pull request by kkofler**
This is a squashed version of @heliocastro's patch set that adds comps support to the hif backend:

Initial support for comps XML parsing. Enable the hif backend to search categories and do a simple mapping of comps categories/groups to PackageKit's static groups.

TODO: support for PackageKit dynamic categories.

Supersedes pull request #20.

### CODE REVIEW

File: backends/hif/pk-backend-hif.c [hunk 1]
```diff
@@ -76,6 +76,147 @@ typedef struct {
...
+	if (comps_data->category_state == COMPS_STATE_GROUP_ID && g_strcmp0 (comps_data->current_query, text) == 0)
+		comps_data->query_match = TRUE;
+
+	if (comps_data->query_match) {
+		if (comps_data->category_state == COMPS_STATE_CATEGORY_GROUPID) {
+			g_debug ("Group: %s", text);
+			g_ptr_array_add (comps_data->groups, g_strdup(text));
+		}
+		if (comps_data->category_state == COMPS_STATE_GROUP_DESCRIPTION) {
+			g_debug ("Description: %s\n", text);
```
Comment (hughsie) [inline backends/hif/pk-backend-hif.c]: No newline for g_debug required.

File: backends/hif/pk-backend-hif.c [hunk 2]
```diff
@@ -76,6 +76,147 @@ typedef struct {
...
+static void
+pk_comps_element_text (GMarkupParseContext *context,
+    const gchar         *text,
+    gsize                text_len,
+    gpointer             user_data,
+    GError             **error)
+{
+	PkCompsData *comps_data = (PkCompsData*) (user_data);
+
+	if (comps_data->category_state == COMPS_STATE_CATEGORY_ID && g_strcmp0 (comps_data->current_query, text) == 0)
```
Comment (hughsie) [inline backends/hif/pk-backend-hif.c]: Two lines please

File: backends/hif/pk-backend-hif.c [hunk 3]
```diff
@@ -1078,6 +1219,214 @@ pk_backend_search_files (PkBackend *backend,
...
+	GMarkupParseContext *context = NULL;
+	PkCompsData *comps_data = (PkCompsData*) (user_data);
+
+	comps_data->category_state = COMPS_STATE_IGNORE;
+
+	context = g_markup_parse_context_new (&parser, 0, comps_data, NULL);
+
+	for (i = 0; i < comps_data->comps->len; i++) {
+		g_debug ("Comps file parsed: %s.", (char*) (g_ptr_array_index (comps_data->comps, i)));
+		if (g_file_get_contents (g_ptr_array_index (comps_data->comps, i), &text, &length, NULL) == FALSE) {
```
Comment (hughsie) [inline backends/hif/pk-backend-hif.c]: !g_file_get_contents rather than == FALSE

File: backends/hif/pk-backend-hif.c [hunk 4]
```diff
@@ -1078,6 +1219,214 @@ pk_backend_search_files (PkBackend *backend,
...
+			g_debug ("Couldn't load XML");
+			return FALSE;
+		} else if (g_markup_parse_context_parse (context, text, length, NULL) == FALSE) {
```
Comment (hughsie) [inline backends/hif/pk-backend-hif.c]: same

**Comment by kkofler** [comment]
Fixed all of the above and 2 more occurrences of `== FALSE` (for a total of 4).

**Comment by kkofler** [comment]
Ping? I addressed all your comments, can this be merged now?

**Comment by hughsie** [comment]
This needs rebasing and further work. Please reopen if you still want to do this work.

**Pull request closed without merging**